File: stack/score-of-parentheses/score-of-parentheses.js
/**
 * @param {string} S
 * @return {number}
 */
var scoreOfParentheses = function (S) {
    let stack = [0];
    for (let s of S) {
        if (s === '(') {
            stack.push(0);
        } else {
            let peek = stack.pop();
            stack[stack.length - 1] += Math.max(peek * 2, 1)
        }
    }

    return stack[0];
};


console.log(scoreOfParentheses('(()(()))'));

File: stack/score-of-parentheses/score_of_parentheses.py
class Solution(object):
    def scoreOfParentheses(self, S):
        stack = [0]

        for x in S:
            if x == '(':
                stack.append(0)
            else:
                peek = stack.pop()
                stack[-1] += max(2 * peek, 1)

        return stack.pop()
